### PR TITLE
Sql support for Compact format [API-367]

### PR DESCRIPTION
**Pull request by sancar**
- [ ] Update reference manual https://docs.hazelcast.com/hazelcast/5.0-snapshot/sql/imap-connector.html

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java [hunk 1]
```diff
@@ -42,6 +43,11 @@ public GenericRecord readGenericRecordOrPrimitive(GenericRecord genericRecord, S
         public void writeFieldFromRecordToWriter(DefaultCompactWriter writer, GenericRecord genericRecord, String fieldName) {
             writer.writeGenericRecord(fieldName, genericRecord.getGenericRecord(fieldName));
         }
+
+        @Override
+        public String readAsJsonFormattedField(AbstractGenericRecord record, String fieldName) {
```
Comment (gierlachg) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java:48]: Is it required for SQL compact support? Maybe it could/should be part of separate PR (there are more changes like this)?

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactContext.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    private final CompactStreamSerializer compactStreamSerializer;
+
+
+    public CompactContext(CompactStreamSerializer compactStreamSerializer) {
+        this.compactStreamSerializer = compactStreamSerializer;
+    }
+
+    @Nullable
+    public Schema getSchema(@Nonnull String typeName) {
```
Comment (gierlachg) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactContext.java:35]: It does not seem to be used 🤔

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java [hunk 3]
```diff
@@ -237,19 +240,24 @@ private Schema getOrReadSchema(ObjectDataInput input, boolean schemaIncludedInBi
...
     }
 
-    private CompactSerializableRegistration getOrCreateRegistration(String className) {
-        return classNameToRegistrationMap.computeIfAbsent(className, s -> {
+    private CompactSerializableRegistration getOrCreateRegistration(String typeName) {
+        return typeNameToRegistrationMap.computeIfAbsent(typeName, s -> {
             Class<?> clazz;
             try {
-                clazz = ClassLoaderUtil.loadClass(classLoader, className);
+                //when the registration does not exist, we tread typeName as className to check if there is a class
```
Comment (gierlachg) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializer.java:258]:
```suggestion
                //when the registration does not exist, we treat typeName as className to check if there is a class
```

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java [hunk 4]
```diff
@@ -48,24 +50,19 @@ private MapSampleMetadataResolver() {
...
-        Object target,
-        boolean key
+            InternalSerializationService ss,
+            JetMapMetadataResolver jetMapMetadataResolver,
+            Object target,
+            boolean key
     ) {
         try {
-            // Convert Portable object to Data to have consistent object fields irrespectively of map's InMemoryFormat.
-            if (target instanceof Portable) {
```
Comment (gierlachg) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:65]: Why it was removed?

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/InternalSerializationService.java [hunk 5]
```diff
@@ -130,4 +132,9 @@
 
     byte getVersion();
 
+    /**
+     * @return compact context
+     */
+    @Nonnull
+    CompactContext getCompactContext();
```
Comment (gierlachg) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/InternalSerializationService.java:139]: It does not seem to be used except for the test variant of  `InternalSerializationService`.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolver.java [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+    @Override
+    public Stream<MappingField> resolveAndValidateFields(boolean isKey, List<MappingField> userFields,
+                                                         Map<String, String> options,
+                                                         InternalSerializationService serializationService) {
+
+        Map<QueryPath, MappingField> userFieldsByPath = extractFields(userFields, isKey);
+        Schema schema = resolveSchema(isKey, options, userFieldsByPath);
+
+        return userFields.isEmpty()
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolver.java:70]: My check was inside resolveSchema. I will check `KvMetadataAvroResolver` and make it similar for the sake of making it similar.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolver.java [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        maybeAddDefaultField(isKey, resolvedFields, fields);
+
+        return new KvMetadata(fields, GenericQueryTargetDescriptor.DEFAULT, new CompactUpsertTargetDescriptor(schema)
+        );
+    }
+
+    @Nonnull
+    private static Schema resolveSchema(boolean isKey, Map<String, String> options,
+                                        Map<QueryPath, MappingField> fieldsByPath) {
+        String typeNameProperty = isKey ? OPTION_KEY_COMPACT_TYPE_NAME : OPTION_VALUE_COMPACT_TYPE_NAME;
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolver.java:197]: No, we are not. Let me add  add checks for it.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvRowProjector.java [hunk 8]
```diff
@@ -92,6 +93,21 @@
         return extractors;
     }
 
+    public Object[] projectData(Data key, Data value) {
+        keyTarget.setTarget(null, key);
```
Comment (sancar) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvRowProjector.java:97]: @gierlachg This does not seem to play well with all the QueryTargets. Some QueryTargets (`AvroQueryTarget`  and `JsonQueryTarget` ) expects the second parameter as null. 
I wonder why there was no test failure. Are tests missing or we never expect `Avro/JsonQueryTarget` here ?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java (hunk 4 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:65]: Not removed but replaced.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactContext.java (hunk 2 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactContext.java:35]: Yes, after refactoring the code it seems it is not needed anymore.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/CompactUpsertTarget.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+import java.time.OffsetDateTime;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.hazelcast.jet.sql.impl.inject.UpsertInjector.FAILING_TOP_LEVEL_INJECTOR;
+
+@NotThreadSafe
+class CompactUpsertTarget implements UpsertTarget {
+
+    private static final Object NOT_SET = new Object();
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/CompactUpsertTarget.java:42]: Do you mean the whole class ? What about MetadataCompactResolver.resolveMetadata ? 
Don't we need to create an CompactUpsertTargetDescriptor from that path ?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlCompactTest.java [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+        assertEquals(row.getObject(0), GenericRecordBuilder
+                .compact(PERSON_ID_TYPE_NAME)
+                .setInt("id", 1).build());
+        assertEquals(row.getObject(1), GenericRecordBuilder
+                .compact(PERSON_TYPE_NAME)
+                .setString("name", "Alice").build());
+    }
+
+    @Test
+    public void test_derivesClassDefinitionFromSchema() throws IOException {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlCompactTest.java:487]: Removing it.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlPortableTest.java [hunk 11]
```diff
@@ -535,11 +534,11 @@ public void test_topLevelFieldExtraction() {
         SqlRow row = rowIterator.next();
         assertFalse(rowIterator.hasNext());
 
-        assertThat(row.<Object>getObject(0)).isEqualToComparingFieldByField(
+        assertEquals(row.<Object>getObject(0),
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlPortableTest.java:537]: It is not returning exact same implementation anymore. AssertEquals is enough from the user perspective. 
I will fix the order of expected and actual.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java (hunk 1 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java]: Removed these changes from the pr. I will send them as a separate one.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/InternalSerializationService.java (hunk 5 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/InternalSerializationService.java:139]: Removing this one as well.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/CompactUpsertTarget.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (gierlachg) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/CompactUpsertTarget.java:42]: I mean the usage of `NOT_SET`. It's there (for `Portable`) to workaround the need to set every field on the record even if it wasn't mapped using `CREATE MAPPING`. Check how is it used.

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/CompactUpsertTarget.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/CompactUpsertTarget.java]: I have removed `schema` from CompactUpserTarget and left only typename. Not_Set is not needed anymore as you said, removed that one as well.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvRowProjector.java [hunk 12]
```diff
@@ -92,6 +93,21 @@
         return extractors;
     }
 
+    public Object[] projectData(Data key, Data value) {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvRowProjector.java]: I have refactored and addded a test to KvRowProjectorTest

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolver.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            List<MappingField> resolvedFields,
+            Map<String, String> options,
+            InternalSerializationService serializationService
+    ) {
+        Map<QueryPath, MappingField> fieldsByPath = extractFields(resolvedFields, isKey);
+
+        String typeNameProperty = isKey ? OPTION_KEY_COMPACT_TYPE_NAME : OPTION_VALUE_COMPACT_TYPE_NAME;
+        String typeName = options.get(typeNameProperty);
+
+        if (typeName == null) {
```
Comment (gierlachg) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolver.java:144]: It should be validated in `resolveAndValidateFields`. Currently, `resolveMetadata` is called after mapping creation, during query execution for each mapping even if it's not used for given query.

**Review by gierlachg — APPROVED** [review]

**Comment by sancar** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java [hunk 14]
```diff
@@ -156,18 +159,36 @@ private static boolean skipMethod(@Nonnull Class<?> clazz, @Nonnull Method metho
...
+     */
+    @Nonnull
+    public static SortedMap<String, QueryDataType> resolveCompact(@Nonnull Schema schema) {
+        SortedMap<String, QueryDataType> fields = new TreeMap<>();
+
+        // Add regular fields.
+        for (String name : schema.getFieldNames()) {
+            FieldType compactType = schema.getField(name).getType();
+            QueryDataType type = resolveType(compactType);
+            fields.putIfAbsent(name, type);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java:182]: Minor: Is it possible to have duplicate field names in the schema?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/extract/GenericQueryTarget.java [hunk 15]
```diff
@@ -101,7 +101,7 @@ private Object prepareTargetForFieldAccess() {
         } else {
             assert serialized != null;
 
-            if (serialized.isPortable()) {
+            if (serialized.isPortable() || serialized.isCompact()) {
                 // Return Portable as Data.
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/extract/GenericQueryTarget.java:105]: Minor: Maybe update this comment as well

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java [hunk 16]
```diff
@@ -48,40 +50,42 @@ private MapSampleMetadataResolver() {
...
+            } else if (target instanceof Portable) {
+                // Convert Portable object to Data to have consistent object fields irrespectively of map's InMemoryFormat.
+                Data data = ss.toData(target);
+                return resolvePortable(ss.getPortableContext().lookupClassDefinition(data), key, jetMapMetadataResolver);
             } else if (target instanceof PortableGenericRecord) {
-                // We get PortableGenericRecord here when in-memory format is OBJECT and
-                // the cluster does not have PortableFactory configuration for this Portable.
-                // PortableGetter can extract fields PortableGenericRecord.
                 return resolvePortable(((PortableGenericRecord) target).getClassDefinition(), key, jetMapMetadataResolver);
+            } else if (target instanceof CompactInternalGenericRecord) {
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:87]: Why it is `PortableGenericRecord` above (instead of `PortableInternalGenericRecord`), but `CompactInternalGenericRecord` here (instead of `CompactGenericRecord`)?

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java [hunk 17]
```diff
@@ -121,16 +125,53 @@ private static MapSampleMetadata resolvePortable(
...
-            jetMapMetadataResolver.resolvePortable(classDef, isKey),
-            new LinkedHashMap<>(fields)
+                GenericQueryTargetDescriptor.DEFAULT,
+                jetMapMetadataResolver.resolvePortable(classDef, isKey),
+                new LinkedHashMap<>(fields)
+        );
+    }
+
+    /**
+     * Resolve metadata from a portable object.
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:135]: Stale documentation. `portable` -> `compact`

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java [hunk 18]
```diff
@@ -121,16 +125,53 @@ private static MapSampleMetadata resolvePortable(
...
+        String topName = isKey ? QueryPath.KEY : QueryPath.VALUE;
+        QueryPath topPath = isKey ? QueryPath.KEY_PATH : QueryPath.VALUE_PATH;
+        // explicitly remove to have the newly-inserted topName at the end
+        fields.remove(topName);
+        fields.put(topName, new MapTableField(topName, QueryDataType.OBJECT, !fields.isEmpty(), topPath));
+
+        return new MapSampleMetadata(
+                GenericQueryTargetDescriptor.DEFAULT,
+                jetMapMetadataResolver.resolveCompact(schema, isKey),
+                new LinkedHashMap<>(fields)
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:167]: Why don't we directly return `fields`?

**Review by mdumandag — COMMENTED** [review]
Looks good overall. Left some comments

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvRowProjector.java [hunk 19]
```diff
@@ -93,8 +94,16 @@
     }
 
     public Object[] project(Object key, Object value) {
-        keyTarget.setTarget(key, null);
-        valueTarget.setTarget(value, null);
+        return project(key, null, value, null);
+    }
+
+    public Object[] project(Data key, Data value) {
+        return project(null, key, null, value);
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvRowProjector.java:101]: No, it will not cause problems. We use this variant for only map connector and Csv/JsonQueryTarget will not be passed there.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlCompactTest.java [hunk 20]
```diff
@@ -0,0 +1,516 @@
...
+        );
+
+        // insert record against new schema/class definition
+        sqlService.execute("SINK INTO " + name + " VALUES (2, 'Bob')");
+
+        // assert both - initial & evolved - records are correctly read
+        assertRowsAnyOrder(
+                "SELECT * FROM " + name,
+                asList(
+                        new Row(1, "Alice"),
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlCompactTest.java:326]: Yes, right. The data is still there.

File: hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java (hunk 14 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java:182]: You can not pass same field name twice in the `CREATE MAPPING`, it is already checked somewhere else. 
`SqlCreateMapping.validate`

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java (hunk 16 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:87]: Looks like a leftover. I will put `CompactGenericRecord` here

File: hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java (hunk 18 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/sql/impl/schema/map/sample/MapSampleMetadataResolver.java:167]: I copied directly from Portable but `fields` is not shared with anyone else. I think we can use it directly. I will change portable as well.

File: hazelcast-sql/src/test/java/com/hazelcast/sql/SqlClientCompactQueryTest.java [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+                .setDateArray("dt", dates)
+                .setTimestampArray("tmstmp", dateTimes)
+                .setTimestampWithTimezoneArray("tmstmptz", offsetDateTimes)
+                .build());
+
+        client.getSql().execute("CREATE MAPPING " + "test2" + '('
+                + "__key INT "
+                + ", b OBJECT "
+                + ", bytes OBJECT"
+                + ", s OBJECT"
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/test/java/com/hazelcast/sql/SqlClientCompactQueryTest.java]: Correcting myself after checking the code. The schema is not created at the MAPPING phase. It is created when `INSERT INTO` is called. At that phase, we have the fields deserialized and we are doing `instanceOf` checks to decide field type of the Compact.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlClientCompactQueryTest.java [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+                + ", '" + OPTION_VALUE_COMPACT_TYPE_NAME + "'='" + "testObject" + '\''
+                + ")"
+        );
+
+        client.getSql().execute("SINK INTO test "
+                        + "( __key, bool, b, s, c, i, l, f, d, st, dl, t, dt, tmstmp, tmstmptz"
+                        + ", booleans, bytes, shorts, chars, ints, longs, floats, doubles, strings,"
+                        + " decimals, times, dates, tmstmps, tmstmptzs)"
+                        + " VALUES (? "
+                        + ",CAST (? AS BOOLEAN)"
```
Comment (ihsandemir) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlClientCompactQueryTest.java:266]: I did not like the way that we have to cast the primitives but not the arrays when using. I was expecting all behave the same way. Can this be a backward compatibility issue when the array types are introduced later on?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolverTest.java [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+import static com.hazelcast.jet.sql.impl.connector.SqlConnector.OPTION_VALUE_COMPACT_TYPE_NAME;
+import static com.hazelcast.jet.sql.impl.connector.map.MetadataCompactResolver.INSTANCE;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@RunWith(JUnitParamsRunner.class)
+public class MetadataCompactResolverTest {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolverTest.java:51]: I don't know the rules of hazelcast-sql module and I just follow what they done. See `MetadataPortableResolverTest` 
@gierlachg may answer better.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlCompactTest.java [hunk 24]
```diff
@@ -0,0 +1,545 @@
...
+import static java.util.Collections.singletonList;
+import static java.util.Spliterator.ORDERED;
+import static java.util.Spliterators.spliteratorUnknownSize;
+import static java.util.stream.StreamSupport.stream;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+public class SqlCompactTest extends SqlTestSupport {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlCompactTest.java:73]: I don't know the rules of hazelcast-sql module and I just follow what they done. See `SqlPortableTest` 
@gierlachg may answer better.

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlCompactTest.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (gierlachg) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/map/SqlCompactTest.java:73]: `SqlTestSupport` provides them, I guess that's enough.
